File: Python3.x/64-Minimum Path Sum.py
# dp
# space: 1
class Solution:
    def minPathSum(self, grid: 'List[List[int]]') -> 'int':
        row, col = len(grid), len(grid[0])
        for i in range(1, col): grid[0][i] += grid[0][i - 1]
        for i in range(1, row): grid[i][0] += grid[i - 1][0]
        for r in range(1, row):
            for c in range(1, col):
                grid[r][c] = min(grid[r-1][c], grid[r][c-1]) + grid[r][c]
        return grid[row-1][col-1]


'''
dp直接在输入的grid矩阵中更新
'''


# dp
# space: row * col
class Solution:
    def minPathSum(self, grid: 'List[List[int]]') -> 'int':
        dp = [[0 for _ in range(len(grid[0]))] for _ in range(len(grid))]
        dp[0][0] = grid[0][0]
        for i in range(1, len(grid)): dp[i][0] = dp[i - 1][0] + grid[i][0]
        for i in range(1, len(grid[0])): dp[0][i] = dp[0][i - 1] + grid[0][i]
        for row in range(1, len(grid)):
            for col in range(1, len(grid[0])):
                dp[row][col] = min(dp[row][col - 1], dp[row - 1][col]) + grid[row][col]
        return dp[-1][-1]

File: C++/64-Minimum_Path_Sum.cpp
// Runtime: 12 ms, faster than 50.16% of C++ online submissions for Minimum Path Sum.
// Memory Usage: 10.7 MB, less than 65.17% of C++ online submissions for Minimum Path Sum.
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        vector<vector<int>> dp(grid.size(), vector<int>(grid[0].size()));
        dp[0][0] = grid[0][0];
        for (int i = 1; i < grid.size(); ++i)
            dp[i][0] = grid[i][0] + dp[i - 1][0];
        for (int i = 1; i < grid[0].size(); ++i)
            dp[0][i] = grid[0][i] + dp[0][i - 1];
        for (int i = 1; i < grid.size(); ++i)
            for (int j = 1; j < grid[0].size(); ++j)
                dp[i][j] = min(dp[i - 1][j], dp[i][j - 1]) + grid[i][j];
        return dp[grid.size() - 1][grid[0].size() - 1];
    }
};


// Runtime: 4 ms, faster than 99.34% of C++ online submissions for Minimum Path Sum.
// Memory Usage: 10.4 MB, less than 100.00% of C++ online submissions for Minimum Path Sum.
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        vector<int> dp(grid[0].size());
        dp[0] = grid[0][0];
        for (int i = 1; i < grid[0].size(); ++i)
            dp[i] = grid[0][i] + dp[i - 1];
        for (int i = 1; i < grid.size(); ++i) { // row
            dp[0] += grid[i][0];
            for (int j = 1; j < grid[0].size(); ++j) // col
                dp[j] = min(dp[j - 1], dp[j]) + grid[i][j];
        }
        return dp[grid[0].size() - 1];
    }
};
